spectator-api/src/main/java/com/netflix/spectator/api/patterns/CardinalityLimiters.java

```diff
@@ -357,4 +377,37 @@ private static class RollupLimiter implements Function<String, String>, Serializ
       return "RollupLimiter(" + state + ")";
     }
   }
+
+  private static class RegisteredNameOrIpLimiter implements Function<String, String>, Serializable {
+
+    private static final long serialVersionUID = 1L;
+
+    //From RFC 3986 3.2.2, we can quickly identify IP literals (other than IPv4) from the first character.
+    private static final Predicate<String> IS_IP_LITERAL = s -> s.startsWith("[");
```

Comment:
One observation is that this matching is inconsistent with the IPv4 check in that the IPv4 will match the pattern anywhere in the string and this would only work if it was at the start. For example:

```
http://127.0.0.1:1234/foo -> is considered an IP
http://[::1]:1234/foo     -> is not because IP is not at the start
```

Should it be `s.contains("[")`?

`IS_IP_V6_LITERAL`? I find the naming to be a bit confusing as "IP" in this context does not include IPv4 addresses. I realize in the future it could be used for future IP versions, but presently it would just be 6.

Reply:
Both good catches, thanks!

The IPv4 pattern should be anchored. The intent is for this limiter to be applicable to RFC 3986 3.2.2 hosts (or the reg_name/host components for various older standards). As noted above, this code is intended to be used on the hot path so I applied a narrow scope to the input to simplify the checks. I'll note that in the javadoc.

Reply:
Actually, I think we should stick with `IS_IP_LITERAL` as that aligns with the name in the specification, even though it does not include IPv4 addresses. The predicate for IPv4 is adjacent and both predicates are scoped to a very small section of code, so I believe the cognitive overhead is low.

Additionally, the versioning in the spec is on the format of the address, not the IP version, so parsers and matcher explicitly don't necessarily require updates when future specs are issued.

For consistency, I've also updated the IPv4 pattern to fully align with the specification.

Reply:
If we are going to anchor both sides of v4, then it should probably be `s.startsWith("[") && s.endsWith("]")`.